source/d1celtilesetframe.cpp

```diff
@@ -158,3 +160,292 @@ void D1CelTilesetFrame::LoadRightTrapezoid(QByteArray &rawData)
+            pixel = this->getPixel(x, y);
+            if (!pixel.isTransparent()) {
+                QMessageBox::critical(nullptr, "Error", "Invalid non-transparent pixel in a Right Triangle frame I.");
+                return pDst;
+            }
+        }
+    }
+
+    for (i = 2; i != MICRO_HEIGHT; i += 2, y--) {
+        // add opaque pixels
+        for (x = 0; x < (MICRO_WIDTH - i); x++) {
+            pixel = this->getPixel(x, y);
+            if (pixel.isTransparent()) {
+                QMessageBox::critical(nullptr, "Error", "Invalid transparent pixel in a Right Triangle frame II.");
+                return pDst;
+            }
+            *pDst = pixel.getPaletteIndex();
+            ++pDst;
+        }
+        pDst += i & 2;
+        // check transparent pixels
+        for (x = MICRO_WIDTH - i; x < MICRO_WIDTH; x++) {
+            pixel = this->getPixel(x, y);
+            if (!pixel.isTransparent()) {
+                QMessageBox::critical(nullptr, "Error", "Invalid non-transparent pixel in a Right Triangle frame II.");
+                return pDst;
+            }
+        }
+    }
+    return pDst;
+}
+
+quint8 *D1CelTilesetFrame::WriteLeftTrapezoid(quint8 *pDst)
+{
+    D1CelPixel pixel;
+    int i, x, y;
+    int length = (MICRO_WIDTH * MICRO_HEIGHT) / 2 + MICRO_HEIGHT * (2 + MICRO_HEIGHT) / 4 + MICRO_HEIGHT / 2;
+
+    // memset(pDst, 0, length);
+    y = MICRO_HEIGHT - 1;
+    for (i = MICRO_HEIGHT - 2; i >= 0; i -= 2, y--) {
+        // check transparent pixels
+        for (x = 0; x < i; x++) {
+            pixel = this->getPixel(x, y);
+            if (!pixel.isTransparent()) {
+                QMessageBox::critical(nullptr, "Error", "Invalid non-transparent pixel in a Left Trapezoid frame I.");
+                return pDst;
+            }
+        }
+        pDst += i & 2;
+        // add opaque pixels
+        for (x = i; x < MICRO_WIDTH; x++) {
+            pixel = this->getPixel(x, y);
+            if (pixel.isTransparent()) {
+                QMessageBox::critical(nullptr, "Error", "Invalid transparent pixel in a Left Trapezoid frame I.");
+                return pDst;
+            }
+            *pDst = pixel.getPaletteIndex();
+            ++pDst;
+        }
+    }
+    // add opaque pixels
+    for (i = MICRO_HEIGHT / 2; i != 0; i--, y--) {
+        for (x = 0; x < MICRO_WIDTH; x++) {
+            pixel = this->getPixel(x, y);
+            if (pixel.isTransparent()) {
+                QMessageBox::critical(nullptr, "Error", "Invalid transparent pixel in a Left Trapezoid frame II.");
+                return pDst;
+            }
+            *pDst = pixel.getPaletteIndex();
+            ++pDst;
+        }
+    }
+    return pDst;
+}
+
+quint8 *D1CelTilesetFrame::WriteRightTrapezoid(quint8 *pDst)
+{
+    D1CelPixel pixel;
+    int i, x, y;
+    int length = (MICRO_WIDTH * MICRO_HEIGHT) / 2 + MICRO_HEIGHT * (2 + MICRO_HEIGHT) / 4 + MICRO_HEIGHT / 2;
+
+    // memset(pDst, 0, length);
```

Comment:
The game does not care whether the memory is cleared. It might be useful if you want to compare the results of two tilesets and you want to reduce the number of differences.

Reply:
In that case lets enable this, it's better to always produce clean dependable data, the game can then figure out if it can ignore things to optimize performance (the editor doesn't have to be particularly well optimized)

Reply:
currently it is not needed, because the caller does this.